src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -0,0 +1,161 @@
+
+    /**
+     * Performs score combination based on input combination technique. Mutates input object by updating combined scores
+     * Main steps we're doing for combination:
+     *  - create map of normalized scores per doc id
+     *  - using normalized scores create another map of combined scores per doc id
+     *  - count max number of hits among sub-queries
+     *  - sort documents by scores and take first "max number" of docs
+     *  - update query search results with normalized scores
+     *  Different score combination techniques are different in step 2, where we create map of "doc id" - "combined score",
+     *  other steps are same for all techniques.
+     * @param queryTopDocs query results that need to be normalized, mutated by method execution
+     * @param scoreCombinationTechnique exact combination technique that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final List<CompoundTopDocs> queryTopDocs, final ScoreCombinationTechnique scoreCombinationTechnique) {
+        // iterate over results from each shard. Every CompoundTopDocs object has results from
+        // multiple sub queries, doc ids may repeat for each sub query results
+        return queryTopDocs.stream()
+            .map(compoundQueryTopDocs -> combineShardScores(scoreCombinationTechnique, compoundQueryTopDocs))
+            .collect(Collectors.toList());
+    }
+
+    private float combineShardScores(ScoreCombinationTechnique scoreCombinationTechnique, CompoundTopDocs compoundQueryTopDocs) {
+        if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+            return ZERO_SCORE;
+        }
+        List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+        // - create map of normalized scores results returned from the single shard
+        Map<Integer, float[]> normalizedScoresPerDoc = getNormalizedScoresPerDocument(topDocsPerSubQuery);
+
+        // - create map of combined scores per doc id
+        Map<Integer, Float> combinedNormalizedScoresByDocId = combineScoresAndGetCombinedNormilizedScoresPerDocument(
+            normalizedScoresPerDoc,
+            scoreCombinationTechnique
+        );
+
+        // - sort documents by scores and take first "max number" of docs
+        // create a priority queue of doc ids that are sorted by their combined scores
+        PriorityQueue<Integer> scoreQueue = getPriorityQueueOfDocIds(normalizedScoresPerDoc, combinedNormalizedScoresByDocId);
+        // store max score to resulting list, call it now as priority queue will change after combining scores
+        float maxScore = combinedNormalizedScoresByDocId.get(scoreQueue.peek());
+
+        // - update query search results with normalized scores
+        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, scoreQueue);
```

Comment:
This method is doing two things: 1. updating queryTopDocs, 2. finding and returning maxScore which violate single responsibility principal. Would there be a way to separate the two responsibility out without compromising efficiency?

Reply:
Agree, it does both, but the alternative is not very efficient, we'll need to iterate over all updated docs and get max scores. This is comparing to approach that I took, it  reuses internals of combiner as we have that max score for update of total hits, and this is without extra result scan.

Reply:
I will leave the decision to you but the rule of thumb that I know is not to increase code complexity for non-critical optimization.